### PR TITLE
Key value store api

### PR DESCRIPTION
**Pull request by KevinDyer**

### CODE REVIEW

File: lib/helpers/key-value-manager.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    clear() {
+      return Promise.resolve()
+      .then(() => {
+        this._store.clear();
+        this.emit('clear');
+      });
+    }
+
+    keys() {
```
Comment (nanpx) [inline lib/helpers/key-value-manager.js]: you can shorten this to

```javascript
return Promise.resolve()
.then(() => Array.from(this._store.keys()));
```

File: lib/helpers/key-value-manager.js [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      .then(() => {
+        const keys = [];
+        for (const key of this._store.keys()) {
+          keys.push(key);
+        }
+        return keys;
+      });
+    }
+
+    values() {
```
Comment (nanpx) [inline lib/helpers/key-value-manager.js]: you can shorten this to

```javascript
return Promise.resolve()
.then(() => Array.from(this._store.values()));
```

File: app/elements/base-key-value/base-key-value-behavior.html [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        this.addSocketEventListener(`${tag} set`, this._boundKvOnSet);
+        this.addSocketEventListener(`${tag} delete`, this._boundKvOnDelete);
+        this.addSocketEventListener(`${tag} clear`, this._boundKvOnClear);
+      }
+    },
+
+    kvSet: function(request) {
+      if (this.tag == null) {
+        return Promise.reject(new Error('no tag specified'));
+      }
```
[the remark this answers is not in the document]
Comment (KevinDyer) [reply app/elements/base-key-value/base-key-value-behavior.html]: __Part 1__: Yarp should definitely make a function
__Part 2__: The reason all this PR is littered with "==" and "!=" instead of "===" and "!==" respectively is due to a neat trick I just learned!
``` javascript
console.log(undefined == null); // true
```
This allows for some nice initialization magic that prevents unnecessary requests to the backend.

File: app/elements/base-key-value/base-key-value-document.html [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        .then(() => {
+          if (this.key != null) {
+            return this.kvSet({key: this.key, value: this.data});
+          }
+        })
+        .then(() => value);
+      }
+    });
+  </script>
+</dom-module>
```
[the remark this answers is not in the document]
Comment (KevinDyer) [reply app/elements/base-key-value/base-key-value-document.html]: You got it baws

File: lib/helpers/key-value-manager.js (hunk 1 above)
Comment (KevinDyer) [reply lib/helpers/key-value-manager.js]: woah! Cool, I'll make this change and the values change

File: app/elements/base-key-value/base-key-value-behavior.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (nanpx) [reply app/elements/base-key-value/base-key-value-behavior.html]: Ahh, got it!
